### PR TITLE
changefeedccl: schema change backfill checkpointing

### PR DESCRIPTION
**Pull request by samiskin**
Previously, initial backfills could checkpoint such that pausing and
resuming a changefeed during the backfill would avoid re-scanning
some amount of already-resolved spans

This did not apply to schema changes though, so if during a schema
change backfill a job had to resume, the entire table(s) would be
scanned from the beginning. The checkpoints emitted on initial backfills
after a resumption also did not preserve old checkpointed spans.

This change both fixes the multiple-resumptions issue as well as adding
checkpointing to schema change backfills.

Release note (performance improvement): changefeed backfills initiated 
by schema changes now periodically checkpoint progress to avoid excessive
re-emitting of already emitted spans.

---

Tl;dr of changes:
- `schemaChangeFrontier`’s `boundaryType/boundaryTime` now represent the latest **non-none** boundary type we’ve seen, rather than being cleared out immediately
- Backfills are now denoted by a `schemaChangeFrontier.BackfillTS()` function that returns the timestamp we expect backfill spans to have, so if you receive a ResolvedSpan with a `timestamp == frontier.BackfillTS()`, it should be a span from a backfill
- The `schemaChangeFrontier` struct on the changeFrontier node is now forwarded to the initialHighWater mark when resumed rather than starting from 0 
- The `schemaChangeFrontier` struct on the changeAggregator nodes now on resume both forward all spans to the initialHighWater but also forward the spans stored in the job progress’ Checkpoint to `initialHighWater.Next()` (since checkpointed spans are defined as spans above the highwater)

### CODE REVIEW

**Review by miretskiy — CHANGES_REQUESTED** [review]
Very nice job on this PR.  I have left quite a few comments -- I think this highlights that the manager of frontier timestamp is *very* tricky.  If you feel adventurous, perhaps you can distill some of those comments and add them to some kind of doc so that this very subtle logic is better documented.

> Reviewed 8 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-:-MmcjEp598b7MzBtnbbk:b-thw2kr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @samiskin)

___
*[-- commits, line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmbylxDDV3eImS15jav):*
I think it's a good catch re not preserving checkpoint during backfill resume.  I would pull that
into separate paragraph.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mm_7_2-3-E_L-2OPBWt:-Mm_7_2-3-E_L-2OPBWu:b-b9tzxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L266)):*
> ```Go
> 	initialHighWater, needsInitialScan := getKVFeedInitialParameters(ca.spec)
> 	if needsInitialScan {
> 		ca.backfillTs = ca.spec.Feed.StatementTime
> ```

This line needs a nice comment.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mm_BXRS6VTXFUqcub9k:-Mm_BXRS6VTXFUqcub9l:b-8s8dmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L531)):*
> ```Go
> 	// timestamp immediately after the boundary
> 	if resolved.BoundaryType == jobspb.ResolvedSpan_BACKFILL {
> 		ca.backfillTs = ca.frontier.boundaryTime.Next()
> ```

I would add a method on schemaChangeFrontier struct to return backfill timestamp.
Move this (nice) comment to that method..
nit: "." at the end of the comment.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 540 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mm_6HcE1TQ5KQXt5gPA:-Mm_6HcE1TQ5KQXt5gPB:b-1wjcez) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L540)):*
> ```Go
> 	checkpointBackfill := ca.spec.JobID != 0 && /* enterprise changefeed */
> 		resolved.Timestamp.Equal(ca.backfillTs) &&
> 		canCheckpointScans(&ca.flowCtx.Cfg.Settings.SV, ca.lastFlush)
> ```

nit: perhaps rename to `canCheckpointBackfill`?
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmbvHNt68fyGixUPXiu:-Mmbsvvs1gNUBbHCjty7:b-9cx9jw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L1122)):*
> ```Go
> 		}
> 
> 		// Recover frontier information from job progress
> ```

I don't think you need to do anything special on the frontier side.
Checkpoint spans are the spans that are ahead of the frontier; another way of saying:  they are ahead
of high watermark we have checkpointed.

Change aggregators get checkpoint information.  They initialize their frontiers based on "Watch" timestamp
which is set to highwatermark.  Good catch regarding not reusing checkpoint information on resume!
So, now that that is fixed, aggregators will advance previously checkpointed spans forward.  Then more spans will
complete backfills.  Eventually, aggregators forward new set of spans that are ahead of their local frontier to 
frontier processor.

It doesn't matter for frontier processor exactly what the frontier state was.  It can even have its cf.frontier timestamp
left at 0 (which is what's happening now).  But the fact is that it will receive spans from aggregators that are ahead of
frontier -- and those are the spans that it will checkpoint.

I could buy an argument that for consistency with aggregator frontiers we should forward cf.frontier to high watermark
(the first loop you have).  But I'm sure we don't need to forward based on checkpoints -- that's already done by aggregators.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mm_DZVv8W3WdQYRbvye:-Mm_DZVv8W3WdQYRbvyf:bmw84oy) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L1337)):*
> ```
> f.js.canCheckpointScans()
> ```
Let's rename `canCheckpointBackfill`
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1663 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mm_Bp4W0-Y99n4C5k8c:-Mm_Bp4W0-Y99n4C5k8d:b5c5chr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L1663)):*
> ```Go
> 	f.Entries(func(s roachpb.Span, ts hlc.Timestamp) span.OpResult {
> 		// When a backfill is occuring, the only spans that are above the highwater
> 		// mark are those that are from the backfill
> ```

Nice comment; thanks.
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 974 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mmck3Sc-vpS8AaME8R9:-Mmbp_1k9xgxp0wPWfd-:b-un665g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_test.go#L974)):*
> ```
> 	skip.UnderRace(t)
> 	skip.UnderShort(t)
> ```

I know you're copying it from the existing checkpointing test.. Just wonder how long does the test take
(skipped under short) and also, can it run/pass under race?
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 1084 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmckTyX6T7dUlMgrT0_:-MmckTyX6T7dUlMgrT0a:bq05czv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_test.go#L1084)):*
> ```Go
> 			haveGaps = true
> 			return true
> 		}
> ```

Pity that we have to duplicate quite a lot of code from initial scan checkpointing test.
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 1198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmclkqK65ABpdI6ldV8:-MmclkqK65ABpdI6ldV9:bdk7vsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_test.go#L1198)):*
> ```Go
> 		t.Fatalf("Found violation of CDC's guarantees: %v", entries)
> 	}
> }
> ```

Very nice test.
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 318 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mm_5YDM4b6bmwugGIvP:-Mm_5YDM4b6bmwugGIvQ:b-7mjdd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L318)):*
> ```
> initial checkpoint
> ```
nit: Update comment?
___
*[pkg/jobs/jobspb/jobs.proto, line 666 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmcYQMYBkW95qskCM7H:-MmcYQMZBCVa_Kt2y3JZ:bgm70lh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/jobs/jobspb/jobs.proto#L666)):*
> ```protobuf
>   message Checkpoint {
>     repeated roachpb.Span spans = 1 [(gogoproto.nullable) = false];
>     util.hlc.Timestamp timestamp = 2 [(gogoproto.nullable) = false];
> ```

I'm not thrilled about the fact that we have to add a timestamp to checkpoint record.
Prior to this change, the meaning of checkpoint was pretty clear: these are the spans that we decided
to checkpoint that are ahead of the changefeed high water mark.

I would like to keep this clarity.

Elsewhere in this PR I suggested that you add a method on schema change frontier BackfillTS.
I think that's all you would need; In changeAggregator, we have the following logic right now (i.e. prior to this PR) that checks
if it's time to checkpoint initial scan:

```
	checkpointBackfill := ca.spec.JobID != 0 && /* enterprise changefeed */
		ca.frontier.Frontier().IsEmpty() && /* initial backfill */
		canCheckpointInitialScan(&ca.flowCtx.Cfg.Settings.SV, ca.lastFlush)
```

This works because during initial scan, the frontier is empty, and we always emit spans with timestamp set to "initialHighwater" (which is set to statementTime) when we do initial scan.

Now, with backfills, we also need to emit those spans *iff* we have reached backfill boundary, and the span timestamps is "backfillTS.Next()".  It appears that we can simply make a helper on schema change frontier:

```
func (f *schemaChangeFrontier) backfillTS() hlc.Timestamp {
    frontier := f.Frontier()
     if frontier.empty() {
        // Handle initial backfill.
        return initialHighWater // you will need to pass this value in when constructing schema change frontier.
     }
    
     restarted := frontier.Equal(initialHighWater)
     backfilling := schemaChangeBoundaryReached() && boundaryType == jobspb.ResolvedSpan_BACKFILL

     if backfilling || restarted {
         // Add that nice comment you have in change aggregator.
         return frontier.Next()
     }
     return hlc.Timestamp{}
}
```

With method along the lines of the above,  you can change the the logic to determine if it's time to checkpoint to be something like:

```
checkpointBackfill := ca.spec.JobID != 0 && /* enterprise changefeed */
		resolved.Timestamp.Equal(ca.frontier.BackfillTS()) &&
		canCheckpointInitialScan(&ca.flowCtx.Cfg.Settings.SV, ca.lastFlush)
```

What happens if you reached backfill boundary, and you've check-pointed some spans, but then changefeed restarts.
I think that's fine: we handle restart case above.  It's true that during restart we don't see "BACKFILL" messages from kv
(because we restart at backfill time + 1); but who cares? It's no different than doing initial scan when we expect to see spans which exactly match the timestamp we return.
___
*[pkg/roachpb/span_group.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mmbo9AY-Cm7eN7oFpnJ:-Mmbo9AY-Cm7eN7oFpnK:ba3eug8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/roachpb/span_group.go#L71)):*
> ```Go
> 	})
> }
> ```

Exported method needs a comment.

**Review by samiskin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71848)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[-- commits, line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmbylxDDV3eImS15jav):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think it's a good catch re not preserving checkpoint during backfill resume.  I would pull that
into separate paragraph.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mm_7_2-3-E_L-2OPBWt:-Mmcu1Ey0leBcP9udN9O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L266)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This line needs a nice comment.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mm_BXRS6VTXFUqcub9k:-MnfUPUIEhRuTKdJ5nl5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L531)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I would add a method on schemaChangeFrontier struct to return backfill timestamp.
Move this (nice) comment to that method..
nit: "." at the end of the comment.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmbvHNt68fyGixUPXiu:-Mmuio8c1Pk4z3Sd1MTX:beg0cdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L1122)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't think you need to do anything special on the frontier side.
Checkpoint spans are the spans that are ahead of the frontier; another way of saying:  they are ahead
of high watermark we have checkpointed.

Change aggregators get checkpoint information.  They initialize their frontiers based on "Watch" timestamp
which is set to highwatermark.  Good catch regarding not reusing checkpoint information on resume!
So, now that that is fixed, aggregators will advance previously checkpointed spans forward.  Then more spans will
complete backfills.  Eventually, aggregators forward new set of spans that are ahead of their local frontier to 
frontier processor.

It doesn't matter for frontier processor exactly what the frontier state was.  It can even have its cf.frontier timestamp
left at 0 (which is what's happening now).  But the fact is that it will receive spans from aggregators that are ahead of
frontier -- and those are the spans that it will checkpoint.

I could buy an argument that for consistency with aggregator frontiers we should forward cf.frontier to high watermark
(the first loop you have).  But I'm sure we don't need to forward based on checkpoints -- that's already done by aggregators.
</blockquote></details>

The checkpoint not being forwarded here just results in the first checkpoint of the second run (after first pause) being inaccurate since it can be emitted prior to the aggregator forwarding the old frontier (that includes the checkpoint from the initial run) in its entirety.  This means progress can go backwards if the changefeed is paused immediately after starting (prior to a full CheckpointFrequency interval after the first checkpoint so that the second checkpoint that includes the full frontier is published), which is gonna be a very rare occasion so its not a big deal, it just required me to change the test to wait for a large enough checkpoint instead of just checking the first one with new data 👍

For the sake of developer ergonomics I very much like having the highwater of the Frontier be consistent between running and being resumed, since now I can rely on it to differentiate initial vs resumed-from-schema-change backfills, and it definitely added some friction working with this to observe the unexpected behaviours that came from  it not being up to date (For example I was watching whether the frontier advanced and after resuming it would be true even when getting spans that should've been earlier/at than the highwater mark, and this was because it was advancing from 0.  Earlier on it was also confusing to see the frontiers of the aggregators and the frontier node be different)
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mm_DZVv8W3WdQYRbvye:-Mmu-u-nEPEEi8jJYTCy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L1337)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> f.js.canCheckpointScans()
> ```
Let's rename `canCheckpointBackfill`
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1663 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mm_Bp4W0-Y99n4C5k8c:-Mmu-udq9nUTVi2u5p3k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L1663)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Nice comment; thanks.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 974 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mmck3Sc-vpS8AaME8R9:-MmtwiROB8Rypw3xr-ua:b-xsh4a9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_test.go#L974)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> 	skip.UnderRace(t)
> 	skip.UnderShort(t)
> ```

I know you're copying it from the existing checkpointing test.. Just wonder how long does the test take
(skipped under short) and also, can it run/pass under race?
</blockquote></details>

The test takes ~16s, and now passes under race after I made the SucceedsSoon calls use atomic
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 1084 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmckTyX6T7dUlMgrT0_:-Mnf_jVu9P1tUik97fNj:b-mqry3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_test.go#L1084)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Pity that we have to duplicate quite a lot of code from initial scan checkpointing test.
</blockquote></details>

Yeah, I tried but it was kind of awkward to pull out common code because of the subtle differences.
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 1198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmclkqK65ABpdI6ldV8:-Mmu-JDz11vvDnS2jWXa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_test.go#L1198)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Very nice test.
</blockquote></details>

Done.
___
*[pkg/jobs/jobspb/jobs.proto, line 666 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmcYQMYBkW95qskCM7H:-MmumJVz9LG9xODx89SJ:b-kasobq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/jobs/jobspb/jobs.proto#L666)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not thrilled about the fact that we have to add a timestamp to checkpoint record.
Prior to this change, the meaning of checkpoint was pretty clear: these are the spans that we decided
to checkpoint that are ahead of the changefeed high water mark.

I would like to keep this clarity.

Elsewhere in this PR I suggested that you add a method on schema change frontier BackfillTS.
I think that's all you would need; In changeAggregator, we have the following logic right now (i.e. prior to this PR) that checks
if it's time to checkpoint initial scan:

```
	checkpointBackfill := ca.spec.JobID != 0 && /* enterprise changefeed */
		ca.frontier.Frontier().IsEmpty() && /* initial backfill */
		canCheckpointInitialScan(&ca.flowCtx.Cfg.Settings.SV, ca.lastFlush)
```

This works because during initial scan, the frontier is empty, and we always emit spans with timestamp set to "initialHighwater" (which is set to statementTime) when we do initial scan.

Now, with backfills, we also need to emit those spans *iff* we have reached backfill boundary, and the span timestamps is "backfillTS.Next()".  It appears that we can simply make a helper on schema change frontier:

```
func (f *schemaChangeFrontier) backfillTS() hlc.Timestamp {
    frontier := f.Frontier()
     if frontier.empty() {
        // Handle initial backfill.
        return initialHighWater // you will need to pass this value in when constructing schema change frontier.
     }
    
     restarted := frontier.Equal(initialHighWater)
     backfilling := schemaChangeBoundaryReached() && boundaryType == jobspb.ResolvedSpan_BACKFILL

     if backfilling || restarted {
         // Add that nice comment you have in change aggregator.
         return frontier.Next()
     }
     return hlc.Timestamp{}
}
```

With method along the lines of the above,  you can change the the logic to determine if it's time to checkpoint to be something like:

```
checkpointBackfill := ca.spec.JobID != 0 && /* enterprise changefeed */
		resolved.Timestamp.Equal(ca.frontier.BackfillTS()) &&
		canCheckpointInitialScan(&ca.flowCtx.Cfg.Settings.SV, ca.lastFlush)
```

What happens if you reached backfill boundary, and you've check-pointed some spans, but then changefeed restarts.
I think that's fine: we handle restart case above.  It's true that during restart we don't see "BACKFILL" messages from kv
(because we restart at backfill time + 1); but who cares? It's no different than doing initial scan when we expect to see spans which exactly match the timestamp we return.
</blockquote></details>

Removed the Checkpoint Timestamp since it would just be a duplicate of initialHighWater and it restricts the definition of checkpoint spans from "spans anywhere over highwater" to "spans at this specific timestamp"👍

---

For `BackfillTS()`, I liked the idea of relying on `boundaryTime` / `initialHighWater` but to implement it I had to be more clear and consistent about the definition of `boundaryTime` and `boundaryType` and our reliance on it.

Currently `boundaryType` + `boundaryTime` mean "The type of boundary of the latest resolvedspan event the frontier has been forwarded to and the time it occurred (unless it's a NONE boundary where the time is just 0)".   This is somewhat confusing since a boundary could still be relevant when you've received a later NONE type event.  For example here, I would receive a BACKFILL boundary event, and then as soon as I receive a NONE-type span for that backfill I'd forget I had a backfill boundary since it gets overwritten with NONE.  non-boundary spans clearing the boundary was unintuitive to me.

There was also that case you mentioned of a table completing its backfill and moving on, receiving NONE events that clear the backfill boundary while the backfill is still going on. 

I instead redefined it so that `boundaryType` + `boundaryTime` meant "The last *non-none* boundary we saw and the timestamp it happened at", that way I can use that information even mid-backfill.  If its the case that we won't read from the schema feed until the backfill is done, then I think this "store the latest non-none boundary" fits perfectly and we wouldn't be able to overwrite a currently relevant boundary.
___
*[pkg/roachpb/span_group.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mmbo9AY-Cm7eN7oFpnJ:-Mmtz9sLBEVZ2H2wG4KE:b-o5fskk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/roachpb/span_group.go#L71)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Exported method needs a comment.
</blockquote></details>

Done.
Also added to tests.

**Review by miretskiy — APPROVED** [review]
Few minor nits.  Feel free to merge once addressed.  
This is pretty awesome.  :lgtm_strong: 

> Reviewed 7 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-:-MngStZzA_ro5gI2Mx0P:bgzfd42)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @samiskin)

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmbvHNt68fyGixUPXiu:-MngJlLcASnIjvSnaosS:bkdsaaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_processors.go#L1122)):*
> The checkpoint not being forwarded here just results in the first checkpoint of the second run (after first pause) being inaccurate since it can be emitted prior to the aggregator forwarding the old frontier (that includes the checkpoint from the initial run) in its entirety. 
> This means progress can go backwards if the changefeed is paused immediately after starting (prior to a full CheckpointFrequency interval after the first checkpoint so that the second checkpoint that includes the full frontier is published), which is gonna be a very rare occasion so its not a big deal, it just required me to change the test to wait for a large enough checkpoint instead of just checking the first one with new data 

Progress timestamp cannot go backwards.  See `highwaterAtStart`.
Or.. Perhaps  by progress you mean "backfill checkpoint spans"; which is what I think you mean.

First, thanks for pushing back on this.  
Secondly, The pattern I think you're concerned about is: we're doing a backfill (or initial scan).  Each aggregator gets its frontier advanced based on checkpoint; However, frontier processor does not... So, it is possible that we could receive new checkpoint from
one (or few) aggregators, and thus frontier processor could update checkpoint record loosing previous progress
from other aggregators it has not heard from; of course "it looses" this information only if we have yet another restart.

All of this points me toward the fact that the 1 line comment you have above should really be expanded with the above explanation.  Let's expand the above comment.  Lets also make sure that we specifcally talk about backfill progress -- *not* highwatermark or anything like that -- because highwatermark definitely cannot regress.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1695 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MngSQXL8RvnNEf3TK0T:-MngSQXL8RvnNEf3TK0U:b-ckp29q) ([raw file](https://github.com/cockroachdb/cockroach/blob/96f84616a072ecfd17095b7ba42afaf31278e343/pkg/ccl/changefeedccl/changefeed_processors.go#L1695)):*
> ```Go
> 
> // BackfillTS returns the timestamp of the incoming spans for an ongoing
> // Backfill (either an Initial Scan backfill or a Schema Change backfill)
> ```

pls add returns empty timestamp when not backfilling.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1700 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MngSUkv26xECUv7M6Rr:-MngSUkv26xECUv7M6Rs:b66jdut) ([raw file](https://github.com/cockroachdb/cockroach/blob/96f84616a072ecfd17095b7ba42afaf31278e343/pkg/ccl/changefeedccl/changefeed_processors.go#L1700)):*
> ```Go
> 
> 	// The scan for the initial backfill results in spans sent at StatementTime,
> 	// which is what initialHighWater is set to when performing an initial scan
> ```

nit: "." end of sentence.
Also :+1: on great comment(s) here.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1706 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MngSber1tqdXYuXmBY5:-MngSber1tqdXYuXmBY6:b-wj6eit) ([raw file](https://github.com/cockroachdb/cockroach/blob/96f84616a072ecfd17095b7ba42afaf31278e343/pkg/ccl/changefeedccl/changefeed_processors.go#L1706)):*
> ```Go
> 
> 	// If the backfill is occuring after any initial scan (non-empty frontier),
> 	// then we can only be in a schema change backfill, where the scan is
> ```

s/we can/it can/?
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 1710 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MngSk2TDphdCmF3eD06:-MngSk2TDphdCmF3eD07:blc0p1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/96f84616a072ecfd17095b7ba42afaf31278e343/pkg/ccl/changefeedccl/changefeed_processors.go#L1710)):*
> ```Go
> 	backfilling := f.boundaryType == jobspb.ResolvedSpan_BACKFILL && frontier.Equal(f.boundaryTime)
> 	// If the schema change backfill was paused and resumed, the initialHighWater
> 	// is read from the job progress and is equal to the old BACKFILL boundary
> ```

Did I say that the comments are great?!  :1st_place_medal: 
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 974 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-Mmck3Sc-vpS8AaME8R9:-Mnfo-qRD9MseOSfRJhw:b-89p12z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/ccl/changefeedccl/changefeed_test.go#L974)):*
<details><summary><i>Previously, samiskin (Shiranka Miskin) wrote…</i></summary><blockquote>

The test takes ~16s, and now passes under race after I made the SucceedsSoon calls use atomic
</blockquote></details>

16 is probably high for race test... Not sure.. Try running multiple runs of race test see if it's stable.
I'm okay if you keep skip under race.
___
*[pkg/jobs/jobspb/jobs.proto, line 666 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71848#-MmcYQMYBkW95qskCM7H:-MnfoNWVDyupbLooIe3e:btsjmha) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b72d583916d2ccbc2571bd9f413b9669d113c8c/pkg/jobs/jobspb/jobs.proto#L666)):*
<details><summary><i>Previously, samiskin (Shiranka Miskin) wrote…</i></summary><blockquote>

Removed the Checkpoint Timestamp since it would just be a duplicate of initialHighWater and it restricts the definition of checkpoint spans from "spans anywhere over highwater" to "spans at this specific timestamp"👍

---

For `BackfillTS()`, I liked the idea of relying on `boundaryTime` / `initialHighWater` but to implement it I had to be more clear and consistent about the definition of `boundaryTime` and `boundaryType` and our reliance on it.

Currently `boundaryType` + `boundaryTime` mean "The type of boundary of the latest resolvedspan event the frontier has been forwarded to and the time it occurred (unless it's a NONE boundary where the time is just 0)".   This is somewhat confusing since a boundary could still be relevant when you've received a later NONE type event.  For example here, I would receive a BACKFILL boundary event, and then as soon as I receive a NONE-type span for that backfill I'd forget I had a backfill boundary since it gets overwritten with NONE.  non-boundary spans clearing the boundary was unintuitive to me.

There was also that case you mentioned of a table completing its backfill and moving on, receiving NONE events that clear the backfill boundary while the backfill is still going on. 

I instead redefined it so that `boundaryType` + `boundaryTime` meant "The last *non-none* boundary we saw and the timestamp it happened at", that way I can use that information even mid-backfill.  If its the case that we won't read from the schema feed until the backfill is done, then I think this "store the latest non-none boundary" fits perfectly and we wouldn't be able to overwrite a currently relevant boundary.
</blockquote></details>

Thanks for sticking w/ me on this one.
